src/java.base/share/classes/jdk/classfile/transforms/CodeLocalsShifter.java

```diff
@@ -40,84 +40,141 @@
 import jdk.classfile.instruction.LocalVariableType;
 
 /**
- *
+ * CodeLocalsShifter is a {@link jdk.classfile.CodeTransform} shifting locals to
+ * new index possitions to avoid conflicts during code injection.
+ * Locals with indexes pointing to the 'this' or method arguments are never shifted.
+ * All locals with indexes pointing beyond the method arguments are re-indexed
+ * in order of appearance.
+ * <p>
+ * A unique index for additional local variable can be manually allocated using
+ * {@link jdk.classfile.transforms.CodeLocalsShifter#addLocal(jdk.classfile.TypeKind)}
+ * at any time before, during, or after the transformation.
+ * <p>
+ * A fork of the actual CodeLocalsShifter can be obtained from
+ * {@link jdk.classfile.transforms.CodeLocalsShifter#fork()}.
+ * Forked CodeLocalsShifter starts with clean locals mapping, however it respects original
+ * method argument locals and avoids conflicts with previously allocated or shifted locals.
+ * Forked CodeLocalsShifter does not dynamically sync with its parent, nor vice versa.
+ * <p>
+ * Sample use in complex transformation:
+ * <p>
+ * {@snippet lang=java :
+ *     var localsShifter = new CodeLocalsShifter(methodModel.flags(), methodModel.methodTypeSymbol());
+ *     methodBuilder.transformCode(codeModel,
+ *          localsShifter
+ *          .andThen((codeBuilder, codeElement) -> {
+ *              ...
+ *              codeBuilder.transform(injectedCodeModel,
+ *                                    localsShifter.fork()
+ *                                    .andThen(otherInjectedCodeTransforms));
+ *              ...
+ *          }));
+ * }
```

Comment:
Sorry, I may be slow but I'm not clear from reading this when a "fork()" would be required.  What's the benefit of a forked copy vs creating a second CodeLocalsShifter?  Is this purely a connivence method? 

If you expand on the rationale for when / why to fork, this would be clearer.

Reply:
Thanks for pointing it out. Yes, the "fork" is not clear and I failed to explain it.
Instead of blurry fork I'll refactor it to expose the two critical numbers:
 - fixed number of method arguments slots (excluded from any shifting)
 - and next free available local index
And fork can be then replaced with creation of a new instance with these two numbers inherited.

Reply:
That makes sense to me.  Thanks for adapting this!

Reply:
Fork is gone.
CodeLocalsShifter let the locals tracking and allocations responsibility on CodeBuilder::allocateLocal.
Instance of CodeLocalsShifter is now needed for the injected code only (so it is not necessary to fork from some top tracking instance).